xarray/backends/rasterio_.py

```diff
@@ -298,6 +300,10 @@ def open_rasterio(filename, parse_coordinates=None, chunks=None, cache=None,
     result = DataArray(data=data, dims=('band', 'y', 'x'),
                        coords=coords, attrs=attrs)
 
+    if mask:
+        for nodataval in attrs.get('nodatavals', ()):
+            result = result.where(result != nodataval)
```

Comment:
The downside of this approach is that unlike how to decode metadata for netCDF files, this won't be lazy.

See `CFMaskCoder` in xarray/coding/variables.py for who we do this with netCDF files. In particular, it probably makes sense to use `lazy_elemwise_func` and to move `nodatavals` from `attrs` to `encoding`.